### DISCUSSION TITLE
Duplicate or Copy Safe

### DISCUSSION THREAD

**Discussion by redboyhun** [pspete/psPAS#427] [Ideas]
Hi guys,
anyone did a script for this task?
I'd like to make a duplicate/copy of a Safe, incl. Safe size, Members and Accounts too.
I need it due to an issue, so I can get rid of the history of the accounts.

What I am not sure and could not test it yet, if the Get-PASSafeMember could be used with Add-PASSafeMember 1-to-1. As I saw the Add command uses switches, but the Get gives a PSCustomObject back, so I probably should make some conversion of the Data.
Same applies for accounts I guess.

Maybe someone made this already and can share his/her solution?

**Comment by pspete** [comment]
Hi @redboyhun 

The only switch parameter declared for `Add-PASSafeMember` is `-UseGen1API`.
A PowerShell object can be used to pipe parameter values into the function; there is no conversion required.

A rough idea for the process to copy safe/copy members/copy accounts follows:

```powershell
$SourceName = "SourceSafe"
$DestinationSafe = "DestinationSafe"

$sourceSafe = Get-PASSafe -SafeName $SourceName

#Create New Safe
$sourceSafe | 
    Select-Object -Property description, managingCPM, numberOfDaysRetention, numberOfVersionsRetention | 
    Add-PASSafe -SafeName $DestinationSafe

$sourceMembers = Get-PASSafeMember -SafeName $SourceName

#Copy Members
$sourceMembers | foreach{

    $PSItem | Select-Object -Property MemberName -ExpandProperty Permissions | 
    Add-PASSafeMember -SafeName $DestinationSafe

}

$sourceAccounts = Get-PASAccount -safeName $SourceName

#Copy Accounts
$sourceAccounts | foreach {

    $PSItem | Select-Object -Property address, userName, platformID, platformAccountProperties | 
    Add-PASAccount -safe $DestinationSafe

}
```

**Comment by begunrom** [comment]
Hi @pspete

Is this supposed to work?
`
$pasAccounts = Get-PASAccount -safeName 'Test' -search 'test' -searchType startswith

$pasAccount | Select-Object -Property address, userName, platformID, platformAccountProperties | Add-PASAccount -safe $DestinationSafe

`
When I run it I get an conversion error.
`

Add-PASAccount : Cannot process argument transformation on parameter 'platformAccountProperties'. Cannot convert value "@{LogonDomain=test.test.org}" to type "System.Collec
tions.Hashtable". Error: "Cannot convert the "@{LogonDomain=test.test.org}" value of type "System.Management.Automation.PSCustomObject" to type "System.Collections.Hashtabl
e"."
At line:1 char:97
+ ... ID, platformAccountProperties | Add-PASAccount -safe $DestinationSafe
+                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (@{address=GRFAS...untProperties=}:PSObject) [Add-PASAccount], ParameterBindingArgumentTransformationException
    + FullyQualifiedErrorId : ParameterArgumentTransformationError,Add-PASAccount`

**Comment by pspete** [comment]
@begunrom - its just a rough idea.
`platformAccountProperties` does need to be a hashtable.
If you find a solution share it here.

**Reply by begunrom** [reply]
@pspete 

I could work around it by converting the PSCustomObject to a Hashtable and then add it to the parameters. It would be handy if the library could do that for me.
`
$pasAccounts = Get-PASAccount -safeName 'Test' -search 'test' -searchType startswith

$platformAccountProperties = @{ }
$pasAccount.platformAccountProperties.psobject.properties | Foreach { $platformAccountProperties[$_.Name] = $_.Value }

$pasAccount | Select-Object -Property address, userName, platformID, remoteMachinesAccess | Add-PASAccount -safe $DestinationSafeName -platformAccountProperties $platformAccountProperties
`
